sources/web/datalab/idleTimeout.ts

```diff
@@ -0,0 +1,256 @@
+/*
+ * Copyright 2017 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+/// <reference path="common.d.ts" />
+
+import childProcess = require('child_process');
+import http = require('http');
+import logging = require('./logging');
+import net = require('net');
+import querystring = require('querystring');
+import url = require('url');
+import userManager = require('./userManager');
+
+const idleCheckIntervalSeconds = 5;    // Seconds to wait between idle checks
+
+let idleTimeoutEnabled = true;  // Allow for explicit disable without clearing other values
+let idleTimeoutSeconds = 0;   // Shutdown after being idle this long; disabled if 0 or NaN
+let lastUserActivity: number;  // The epoch, in seconds, of the last user activity
+let shutdownCommand = '';
+
+export function initAndStart(appSettings: common.Settings) {
+  init(appSettings);
+  reset();
+  startChecker();
+}
+
+export function init(appSettings: common.Settings) {
+  shutdownCommand = appSettings.idleTimeoutShutdownCommand;
+  if (!shutdownCommand) {
+    shutdownCommand = process.env.DATALAB_SHUTDOWN_COMMAND;
+  }
+  if (shutdownCommand) {
+    idleTimeoutSeconds = appSettings.idleTimeoutSeconds;
+    logging.getLogger().debug('idleTimeoutSeconds from settings: ' + idleTimeoutSeconds);
+    if (!idleTimeoutSeconds) {
+      idleTimeoutSeconds = parseInterval(process.env.DATALAB_IDLE_TIMEOUT);
+    }
+    logging.getLogger().debug('Idle timeout set to ' + idleTimeoutSeconds + ' seconds');
+    if (! (idleTimeoutSeconds > 0)) {
+      logging.getLogger().debug('No idle timeout value, idle timeout is disabled');
+    }
+  } else {
+    idleTimeoutSeconds = NaN;   // No shutdown command available
+    logging.getLogger().debug('No shutdown command, idle timeout is disabled');
+  }
+}
+
+// Parse a string like '30s' or '1h 15m 5s' and return seconds.
+// If no input string, or unrecognized stuff, returns NaN.
+function parseInterval(str: string) {
```

Comment:
Sounds reasonable. What are the rules regarding adding new packages to what we ship? Licensing guidelines?